### ISSUE TITLE
'Files in use' tracking incredibly slow with large numbers of (nested) folders

### ISSUE THREAD

**Issue by MasonD** [silverstripe/silverstripe-asset-admin#931]
The `filesInUseCount` field of the `Folder` graphql type eventually ends up calling `nestedFolderIDs`, which is incredibly slow with large numbers of folders (in fact it runs approximately one SQL query per folder).

On a project with ~26000 folders, the vast majority of which live somewher under a single top-level folder, we see response times for any graphql query that fetches this folder of upwards of two minutes. Patching `filesInUseCount` to always immediately return an empty DataList speeds up the slow sections of our asset admin by 500 times.

If it's not feasible to increase the performance of `nestedFolderIDs`, perhaps a configuration option to disable `filesInUseCount` is necessary, so large projects can feasibly use the asset admin. 

# Sub tasks
* [x] Add a state to track confirmation modal on front end
* [x] Display modal to confirm deletion of files
* [x] Create GraphQL end point for retrieving file usage for a list of files
* [x] Fetch file usage when display delete confirmation modal
* [x] Adjust confirmation message based on file usage
* [x] Move the delete file logic and hook it into the confirmation modal
* [x] Display loading indicator while the file usage data is being fetch
* [x] Add unit test to test state reducers
* [x] Add test to confirmation modal component
* [x] Update behat test
* [x] Write some unit test for the new GraphQL endpoint
* [x] Stop fetching fileInUseCount on regular readFiles request
* [ ] ~~Stretch goal: Use toast notifications instead of alert boxes~~
* [x] Apply the same logic to the regular single file delete.

# Pull requests
* [x] https://github.com/silverstripe/silverstripe-asset-admin/pull/999

**Comment by dnsl48** [comment]
Thank you for reporting this and taking time to analyse the problem and provide a good detailed explanation of the problem.

> Perhaps a configuration option to disable filesInUseCount is necessary

Yes, this definitely feels like a non essential feature

> On a project with ~26000 folders ... speeds up the slow sections of our asset admin by 500 times

Do you remember approx numbers in seconds/minutes/hours?

**Comment by MasonD** [comment]
No problem! It was a big issue affecting the usability of our assets since we upgraded to silverstripe 4, so getting it fixed for ourselves ASAP was a priority.

I can't recall the numbers for the 500 times figure I quoted, which I believe was for my local development environment, but testing again on a local development environment with a database on the same machine as the server, response times for requests containing this monolithic folder take 16s, and removing filesInUseCount reduces the time to 220ms. As I said in the initial post, on our production environment we observed response times upwards of two minutes in some cases. This has fallen down to 4.9s after removing filesInUseCount.

**Comment by dnsl48** [comment]
> so getting it fixed for ourselves ASAP was a priority

did you find a workaround for that?

**Comment by MasonD** [comment]
Not something that'd be amenable to inclusion upstream. I just patched `AssetAdminFile::getFilesInUse()` to return an empty `ArrayList`.

**Comment by dnsl48** [comment]
Reclassifying as a bug since it makes UI unusable.

**Comment by maxime-rainville** [comment]
At first glance, it looks like the only things this field is used for is to warn the user when they try to delete a directory with files in it.

https://github.com/silverstripe/silverstripe-asset-admin/blob/467f77ea830fa4a13a38800957710502f76d992f/client/src/constants/index.js#L25-L40

We could fix this by:
* Waiting until the user actually wants to delete the folder to do this check.
* Retrieve a simpler boolean value of `hasFileInUse` rather than counting all the files.
* Both of the above.

**Comment by maxime-rainville** [comment]
I've bee playing with this a bit. I'll write down where I got to, mainly so I can get back to it on Monday.

# Core problems
When we fetch the list of folders and files, we request an attribute called `filesInUseCount` for folders and `inUseCount` for regular files. This can be quite computational intensive to generate because it involves going through a big list of queries to see if a folder contain a files that is use somewhere.

Those values are actually only needed when you display a confirmation message. Based on the values you'll either be blocked from deleting folders altogether or you'll get a different error message.

# Proposed fix

Basically only retrieve these values when they are needed, basically when you try to bulk delete files and folders.

The logic for displaying this message is actually in the ironically name `asset-admin/client/src/constants/index.js` file which contains a list of possible Bulk Action. This file looks like it was originally there to hold a short list of possible BULK_ACTIONS with some config options. In practice most of the file is dedicated to a fancy `confirm` function to run when the user wants to delete files/folders.

This BULK_ACTIONS list of bulk actions is used by `Gallery.renderBulkActions()`. My thinking at this stage is moved this confimation logic to `Gallery` and use the `@silverstripe/react-confirm` instead.

On the PHP side, I'm thinking let's create a new GraphQL query that recieves a list of Folder and File IDs. This query will let the front end know if it's safe to delete the files.

# Some problems I'm having

A downside of the `@silverstripe/react-confirm` confirm method is that it seems to operate outside our regular Apollo/Injector context. So it's difficult to run a GraphQL query in that Modal with our usual bags of tricks.

I could just do a manual request using the Apollo client, but that would be less then ideal, because that request could theoretically take a long time if you have lots of files and folders. So I would need to display some loading animation before showing calling confirm.

Alternatively, I guess I could forgo using  `@silverstripe/react-confirm` and add my own dedicated bulk delete modal component and hook into the redux store.

**Comment by maxime-rainville** [comment]
Finally managed to get a start on this. I added some sub tasks and created a draft PR for reference https://github.com/silverstripe/silverstripe-asset-admin/issues/931

@silverstripeux I'm replacing the native JS confirmation message with a modal message. I'm also adding a graphQL endpoint that can return me information about the file usage. This does create some interesting opportunities. e.g.: Tell the user which files are currently in-use amongst the one they're thinking of deleting or even tell them where those files are being used.

Probably worth having a chat about what we would like to do before I get too far down a given track.

**Comment by clarkepaul** [comment]
@maxime-rainville that sounds awesome. 
I would really like some report or way to see "Files which aren't used". Otherwise, the main case is as you've mentioned with deleting.

**Comment by maxime-rainville** [comment]
I've got this thing working. It's very rough around the edges, but it shows where I'm going with this.

There's still a lot of cleaning up to do, but I could use some feedback on the general approach I've taken ... both on the design and implementation.

Here's what this thing looks like https://youtu.be/WxE-oFGRvAc

**Comment by maxime-rainville** [comment]
Had a look at adding toast notifications, but we would need this issue to get over the line before we can do it https://github.com/silverstripe/silverstripe-admin/issues/504

**Comment by maxime-rainville** [comment]
This issue has been worked on as part of a larger initiative focused on reducing the experience debt of the SilverStripe CMS.

Thank you for taking the time to raise this issue and/ or provide feedback supporting the implementation.

This is an initiative from the teams maintaining the commercially supported module list. If you're interested to read more about this initiative and keep up with our progress have a look at our [blog post on experience debt](https://www.silverstripe.org/blog/weve-got-experience-debt-in-our-sights).
